File: packages/server-go/internal/store/lexorank.go
package store

import (
	"strings"
)

type RankItem struct {
	ID   string
	Rank string
}

type RankResult struct {
	ID      string
	NewRank string
}

func GenerateInitialRank() string {
	return "0|hzzzzz"
}

func GenerateRankBetween(before, after string) string {
	var rankA, rankB string
	if before != "" {
		_, rankA = splitRank(before)
	}
	if after != "" {
		_, rankB = splitRank(after)
	}
	return "0|" + midpoint(rankA, rankB)
}

func Rebalance(items []RankItem) []RankResult {
	n := len(items)
	if n == 0 {
		return nil
	}

	const rankLength = 6
	results := make([]RankResult, n)
	for i, item := range items {
		fraction := float64(i+1) / float64(n+1)
		rankChars := make([]byte, rankLength)
		frac := fraction
		for d := 0; d < rankLength; d++ {
			frac *= 26
			digit := int(frac)
			rankChars[d] = byte('a' + digit)
			frac -= float64(digit)
		}
		results[i] = RankResult{ID: item.ID, NewRank: "0|" + string(rankChars)}
	}
	return results
}

func CompareLexorank(a, b string) int {
	if a < b {
		return -1
	}
	if a > b {
		return 1
	}
	return 0
}

func midpoint(a, b string) string {
	// Normalise to equal length (at least 1)
	maxLen := len(a)
	if len(b) > maxLen {
		maxLen = len(b)
	}
	if maxLen == 0 {
		maxLen = 1
	}

	// a empty → minimum, b empty → maximum (matching TS semantics)
	var aa string
	if a != "" {
		aa = padRight(a, maxLen)
	} else {
		aa = strings.Repeat("a", maxLen)
	}

	aDigits := make([]int, maxLen)
	for i := 0; i < maxLen; i++ {
		aDigits[i] = int(aa[i] - 'a')
	}

	bDigits := make([]int, maxLen)
	if b != "" {
		bb := padRight(b, maxLen)
		for i := 0; i < maxLen; i++ {
			bDigits[i] = int(bb[i] - 'a')
		}
	} else {
		for i := 0; i < maxLen; i++ {
			bDigits[i] = 25 // 'z' = max
		}
	}

	// Sum digits (base-26)
	sum := make([]int, maxLen)
	carry := 0
	for i := maxLen - 1; i >= 0; i-- {
		s := aDigits[i] + bDigits[i] + carry
		sum[i] = s % 26
		carry = s / 26
	}

	// Divide by 2 to get midpoint, propagating carry from sum
	remainder := carry
	mid := make([]byte, maxLen)
	for i := 0; i < maxLen; i++ {
		cur := remainder*26 + sum[i]
		mid[i] = byte('a' + cur/2)
		remainder = cur % 2
	}

	// Extra precision digit if remainder
	if remainder > 0 {
		mid = append(mid, byte('a'+13)) // 'm'
	}

	result := strings.TrimRight(string(mid), "a")
	if result == "" {
		result = "a"
	}

	// Safety: if result equals either input, extend
	if result == aa || (b != "" && result == padRight(b, maxLen)) {
		result += string(byte('a' + 13))
	}

	return result
}

func splitRank(rank string) (string, string) {
	parts := strings.SplitN(rank, "|", 2)
	if len(parts) != 2 {
		return "0", rank
	}
	return parts[0], parts[1]
}

func padRight(s string, length int) string {
	for len(s) < length {
		s += "a"
	}
	return s
}

File: packages/server/src/lexorank.ts
/**
 * Lexorank — lexicographic ranking for sortable lists.
 *
 * Format: `bucket|rank` where bucket is a single digit (0-2, currently fixed
 * to 0) and rank is a lowercase-alpha string ordered lexicographically.
 */

const BUCKET = '0';
const MIN_CHAR = 'a';               // code 97
const MAX_CHAR = 'z';               // code 122
const ALPHABET_SIZE = 26;           // a-z
const RANK_LENGTH = 6;              // default rank width

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Pad `s` on the right with MIN_CHAR to length `len`. */
function padRight(s: string, len: number): string {
  while (s.length < len) s += MIN_CHAR;
  return s;
}

/**
 * Compute a midpoint rank string between two rank strings `a` and `b`.
 *
 * - `a` < `b` lexicographically (caller must guarantee).
 * - If `a` is empty it is treated as the absolute minimum.
 * - If `b` is empty it is treated as the absolute maximum.
 * - When `a` and `b` are adjacent the result is longer (one extra char).
 */
export function midpoint(a: string, b: string): string {
  // Normalise to equal length
  const maxLen = Math.max(a.length, b.length, 1);
  const aa = a ? padRight(a, maxLen) : MIN_CHAR.repeat(maxLen);
  let bb = b ? padRight(b, maxLen) : '';

  // Convert rank strings to digit arrays (base-26)
  const digitsA = Array.from(aa, (ch) => ch.charCodeAt(0) - 97);
  const digitsB = bb
    ? Array.from(bb, (ch) => ch.charCodeAt(0) - 97)
    : new Array(maxLen).fill(ALPHABET_SIZE - 1); // 'z' = max valid char

  // Sum digits as a big base-26 number, then halve.
  // We work with integers to avoid floating-point drift.
  let carry = 0;
  const sum: number[] = [];
  for (let i = maxLen - 1; i >= 0; i--) {
    const s = digitsA[i] + digitsB[i] + carry;
    carry = Math.floor(s / ALPHABET_SIZE);
    sum.unshift(s % ALPHABET_SIZE);
  }
  // carry is at most 1 here (since both values < base^len)

  // Divide the sum by 2 to get the midpoint.
  let remainder = carry; // propagate leading carry into division
  const mid: number[] = [];
  for (const d of sum) {
    const cur = remainder * ALPHABET_SIZE + d;
    mid.push(Math.floor(cur / 2));
    remainder = cur % 2;
  }

  // If there is a remainder we need one extra digit of precision.
  if (remainder) {
    mid.push(Math.floor(ALPHABET_SIZE / 2)); // 'm'
  }

  let result = mid.map((d) => String.fromCharCode(d + 97)).join('');

  // Trim trailing MIN_CHARs but keep at least one char
  result = result.replace(/a+$/, '') || MIN_CHAR;

  // Safety: if result equals a or b we extend by one char
  if (result === aa || result === (bb || '')) {
    result += String.fromCharCode(97 + Math.floor(ALPHABET_SIZE / 2));
  }

  return result;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Return a seed rank suitable for the very first item: `0|hzzzzz`. */
export function generateInitialRank(): string {
  return `${BUCKET}|hzzzzz`;
}

/**
 * Generate a rank string between two existing lexoranks.
 *
 * - `before == null` → insert at the very beginning.
 * - `after  == null` → insert at the very end.
 * - Both provided   → insert between them.
 */
export function generateRankBetween(
  before: string | null,
  after: string | null,
): string {
  const rankA = before ? before.split('|')[1] : '';
  const rankB = after ? after.split('|')[1] : '';
  return `${BUCKET}|${midpoint(rankA, rankB)}`;
}

/**
 * Evenly redistribute lexorank values across `items` (order preserved).
 * Useful when the gap between adjacent ranks is exhausted.
 */
export function rebalance(
  items: Array<{ id: string }>,
): Array<{ id: string; position: string }> {
  if (items.length === 0) return [];

  const results: Array<{ id: string; position: string }> = [];

  for (let i = 0; i < items.length; i++) {
    // Treat index as a fraction of the total range and convert to base-26.
    const fraction = (i + 1) / (items.length + 1);
    const rankChars: string[] = [];
    let frac = fraction;
    for (let d = 0; d < RANK_LENGTH; d++) {
      frac *= ALPHABET_SIZE;
      const digit = Math.floor(frac);
      rankChars.push(String.fromCharCode(97 + digit));
      frac -= digit;
    }
    const rank = rankChars.join('');
    results.push({ id: items[i].id, position: `${BUCKET}|${rank}` });
  }

  return results;
}

/**
 * Comparator for two lexorank strings — use with `Array.prototype.sort`.
 */
export function compareLexorank(a: string, b: string): number {
  if (a < b) return -1;
  if (a > b) return 1;
  return 0;
}
